Substitute KVars in constraints for PLE

If the LHS of a constraint has KVars, PLE ignores it. This is unfortunate because unresolved constraints can have KVars in this position.

It should be feasible to change liquid-fixpoint so it replaces these KVars with their best known solutions before feeding them to PLE. In this way, PLE has a chance to discover unfoldings that are needed by these constraints.

There is a fine example of this bug at work in https://github.com/ucsd-progsys/liquidhaskell/issues/2582

--------
### Implementation notes

There is a function `lhsPred` that converts the LHS of a constrain into a predicate with no KVars. We could try assembling a similar function that takes the bindings of a constraint and produces the same bindings with the KVars substituted.

I have an implementation of the KVars substitution in ff7857aea. This is not enough though, because PLE needs to be extended to handle expressions with existentials. Here's the design I'm considering:

1. When PLE first picks a set of bindings in which to look for unfoldings, rename all the existential bindings so they use fresh symbols.
2. When analyzing a candidate expression in search for unfoldings, have PLE remember the existential bindings that it has encountered when traversing the expression in some form of a stack.
3. When PLE finds a new unfolding to make, record together with the new equality, the current stack of existential bindings.
4. When putting the new equalities in the SMT solver environment, use the stacks of existential bindings attached to them, to identify the original existential expressions. Say the original expression was of the form `Ex (p /\ Ey q)`, the new assertion for the predicate environment should be of the form
```
Ex (p /\ new_equalities_under_x /\
    Ey (q /\ new_equalities_under_x_and_y)
   )
```
5. When putting the new equalities back into existing constraints, they should be inserted in constraint environments with the same form as in (4).

In (4) and (5), predicates `p` and `q` might need to be sent multiple times to the SMT solver, we might try simplifying the existential bodies to reduce their size if it turns out to be a problem.

The stack in the design could be avoided. But the removal would cause `p` and `q` to be inserted multiple times in constraints, in cases where they could be inserted only once by properly grouping the equalities they induced.

Here is a second attempt to show that unfoldings are needed in the Kvar solutions:
<details>

```Haskell
{-@ LIQUID "--ple" @-}
module Test where

import Data.Set as Set

{-@ data State s a <p :: s -> Bool, q :: s -> a -> s -> Bool>
      = State (x:s<p> -> (a, s)<\w -> {v:s<q x w> | true}>)  @-}
data State s a  = State {runState :: s -> (a, s)}

{-@ runState :: forall <p :: s -> Bool, q :: s -> a -> s -> Bool>.
                State <p, q> s a -> x:s<p> -> (a, s)<\w -> {v:s<q x w> | true}> @-}

instance Functor (State s) where
{-@
instance Functor (State s) where
  fmap :: forall < p  :: s -> Bool
                 , q :: s -> a -> s -> Bool
                 , q2 :: s -> a -> s -> Bool
                 , pa :: a -> Bool
                 , pb :: b -> Bool
                 >.
            {x::a<pa>, s0::s<p>, y::b<pb> |- s<q s0 x> <: s<q2 s0 y>}
            (a<pa> -> b<pb>)
         -> State <p, q> s a<pa>
         -> State <p, q2> s b<pb>
@-}
  fmap f x = State $ \s -> let (y, s') = runState x s in (f y, s')


instance Applicative (State s) where
{-@
instance Applicative (State s) where
  pure :: forall <p :: s -> Bool>.
                x:a -> State <p, \w0 y -> {w1:s<p> | w0 == w1 && y == x}> s a
@-}
  pure x = State $ \s -> (x, s)
  mf <*> mx = State $ \s ->
    let (f, s1) = runState mf s
        (x, s2) = runState mx s1
    in (f x, s2)

instance Monad (State s) where
{-@
instance Monad (State s) where
  >>= :: forall < p  :: s -> Bool
                    , p2 :: a -> s -> Bool
                    , r  :: a -> Bool
                    , q1 :: s -> a -> s -> Bool
                    , q2 :: a -> s -> b -> s -> Bool
                    , q  :: s -> b -> s -> Bool>.
            {x::a<r>, w::s<p>|- s<q1 w x> <: s<p2 x>}
            {w::s<p>, x::a, w1::s<q1 w x>, y::b |- s<q2 x w1 y> <: s<q w y>}
            {x::a, w::s, w2::s<q1 w x>|- {v:a | v = x} <: a<r>}
            State <p, q1> s a
         -> (x:a<r> -> State <{v:s<p2 x> | true}, \w1 y -> {v:s<q2 x w1 y> | true}> s b)
         -> State <p, q> s b ;
  >>  :: forall < p   :: s -> Bool
                    , p2  :: s -> Bool
                    , q1 :: s -> a -> s -> Bool
                    , q2 :: s -> b -> s -> Bool
                    , q :: s -> b -> s -> Bool>.
            {x::a, w::s<p>|- s<q1 w x> <: s<p2>}
            {w::s<p>, w2::s<p2>, x::b, y::a |- s<q2 w2 x> <: s<q w x>}
            State <p, q1> s a
         -> State <p2, q2> s b
         -> State <p, q> s b  ;
  return :: forall <p :: s -> Bool>.
                x:a -> State <p, \w0 y -> {w1:s<p> | w0 == w1 && y == x}> s a
@-}
  (State g) >>= f = State $ \x -> case g x of {(y, s) -> (runState (f y)) s}
  (State g) >>  f = State $ \x -> case g x of {(y, s) -> (runState f    ) s}
  return w      = State $ \s -> (w, s)

{-@
get :: forall <p :: s -> Bool>.
   State <p, {\s0 v s -> v = s && s = s0}> s s<p>
@-}
get :: State s s
get = State $ \s -> (s, s)

{-@
put :: forall <p :: s -> Bool>.
   s0:s -> State <p, {\_ v s -> s = s0}> s ()
@-}
put :: s -> State s ()
put s = State $ \_ -> ((), s)

{-@ reflect gt0 @-}
gt0 :: Int -> Bool
gt0 x = x > 0

{-@
stFun
  :: i:Int
  -> State
       <{\m0 -> gt0 i}
       , {\m0 v m -> v > 0 }>
       _ _
@-}
stFun :: Int -> State Int Int
stFun i = do
  put i
  get
```
</details>

The error message is
```
Test.hs:110:1: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : GHC.Types.Int
    .
    is not a subtype of the required type
      VV : {VV##2013 : GHC.Types.Int | Test.gt0 ?a}
    .
    in the context
      ?a : GHC.Types.Int
    Constraint id 4
    |
110 | stFun i = do
    | ^^^^^^^^^^^^...
```

The constraint is:
```
constraint:
  lhs {VV : int | [$k_##1822[VV:=VV][VV##1821:=VV][VV##2013:=VV][VV##F:=VV][lq_tmp$db##65:=lq_anf$##d1cb][lq_tmp$db##66:=lq_anf$##d1cc][lq_tmp$x##1823:=lq_tmp$x##1761][lq_tmp$x##1824:=lq_tmp$x##1762][lq_tmp$x##2014:=VV]]}
  rhs {VV : int | [(lq_tmp$x##1762 > 0)]}
  id 4 tag [26]
  // META constraint id 4 : Test.hs:(110,1)-(112,5)
  environment:
    lq_anf$##d1cb : {lq_tmp$x : (Test.State int Tuple0) | []}
    lq_anf$##d1cc : {lq_tmp$x : (Test.State int int) | []}
    lq_tmp$x##1761 : {VV : int | []}
    lq_tmp$x##1762 : {VV : int | []}
```

And `$k_##1822` is a large expression that contains `gt0 i##aS7`.

Unfolding manually `gt0` makes verification pass:
```diff
-       <{\m0 -> gt0 i}
+       <{\m0 -> i > 0}
```

--- a/src/Language/Fixpoint/Smt/Interface.hs
+++ b/src/Language/Fixpoint/Smt/Interface.hs
@@ -106,6 +106,7 @@ import qualified SMTLIB.Backends
 import qualified SMTLIB.Backends.Process as Process
 import qualified Language.Fixpoint.Conditional.Z3 as Conditional.Z3
 import Control.Concurrent.Async (async)
+import GHC.Stack (HasCallStack)
 
 {-
 runFile f
@@ -121,19 +122,22 @@ runCommands cmds
        return zs
 -}
 
-checkValidWithContext :: [(Symbol, Sort)] -> Expr -> Expr -> SmtM Bool
+checkValidWithContext
+  :: HasCallStack => [(Symbol, Sort)] -> Expr -> Expr -> SmtM Bool
 checkValidWithContext xts p q =
   smtBracket "checkValidWithContext" $
     checkValid' xts p q
 
 -- | type ClosedPred E = {v:Pred | subset (vars v) (keys E) }
 -- checkValid :: e:Env -> ClosedPred e -> ClosedPred e -> IO Bool
-checkValid :: Config -> FilePath -> [(Symbol, Sort)] -> Expr -> Expr -> IO Bool
+checkValid
+  :: HasCallStack
+  => Config -> FilePath -> [(Symbol, Sort)] -> Expr -> Expr -> IO Bool
 checkValid cfg f xts p q = do
   me <- makeContext cfg f
   evalStateT (checkValid' xts p q) me
 
-checkValid' :: [(Symbol, Sort)] -> Expr -> Expr -> SmtM Bool
+checkValid' :: HasCallStack => [(Symbol, Sort)] -> Expr -> Expr -> SmtM Bool
 checkValid' xts p q = do
   smtDecls xts
   smtAssertDecl $ pAnd [p, PNot q]
@@ -180,7 +184,7 @@ commandRaw ctxLog ctxSolver ctxVerbose cmdBS = do
 
 --------------------------------------------------------------------------------
 {-# SCC command #-}
-command  :: Command -> SmtM Response
+command  :: HasCallStack => Command -> SmtM Response
 --------------------------------------------------------------------------------
 command !cmd       = do
   -- whenLoud $ do LTIO.appendFile debugFile (s <> "\n")
@@ -429,7 +433,7 @@ smtAssert p = interact' (Assert Nothing p)
 
 -- the following three functions will emit additional `apply`,
 -- `coerce`, and `lambda` symbols for fresh function sorts as needed
-smtAssertDecl :: Expr -> SmtM ()
+smtAssertDecl :: HasCallStack => Expr -> SmtM ()
 smtAssertDecl p = interactDecl' (Assert Nothing p)
 
 smtDefineEqn :: Equation -> SmtM ()
@@ -453,7 +457,7 @@ smtAssertAxiom p  = interact' (AssertAx p)
 smtDistinct :: [Expr] -> SmtM ()
 smtDistinct az = interact' (Distinct az)
 
-smtCheckUnsat :: SmtM Bool
+smtCheckUnsat :: HasCallStack => SmtM Bool
 smtCheckUnsat = respSat <$> command CheckSat
 
 smtBracketAt :: SrcSpan -> String -> SmtM a -> SmtM a
@@ -479,7 +483,7 @@ smtBracket _msg a = do
         , ctxIxs = is}
   return r
 
-respSat :: Response -> Bool
+respSat :: HasCallStack => Response -> Bool
 respSat Unsat   = True
 respSat Sat     = False
 respSat Unknown = False
@@ -490,7 +494,7 @@ interact' cmd  = void $ command cmd
 
 -- | a variant of `interact'` which also emits fresh
 --   `apply`, `coerce`, and `lambda` symbols
-interactDecl' :: Command -> SmtM ()
+interactDecl' :: HasCallStack => Command -> SmtM ()
 interactDecl' cmd  = do
   cmdBS <- liftSym $ runSmt2 cmd
   ctx <- get

--- a/src/Language/Fixpoint/Solver/Common.hs
+++ b/src/Language/Fixpoint/Solver/Common.hs
@@ -10,11 +10,12 @@ import Language.Fixpoint.Types
 import Language.Fixpoint.Types.Visitor (kvarsExpr)
 import Language.Fixpoint.Defunctionalize (defuncAny)
 import Language.Fixpoint.SortCheck (ElabParam(..), elaborate)
+import GHC.Stack (HasCallStack)
 
 mytracepp :: (PPrint a) => String -> a -> a
 mytracepp = notracepp
 
-askSMT :: Config -> [(Symbol, Sort)] -> Expr -> SmtM Bool
+askSMT :: HasCallStack => Config -> [(Symbol, Sort)] -> Expr -> SmtM Bool
 askSMT cfg xs e
 --   | isContraPred e  = return False
   | isTautoPred  e     = return True
@@ -24,7 +25,7 @@ askSMT cfg xs e
          checkValidWithContext xs PTrue e'
   | otherwise          = return False
 
-toSMT :: String -> Config -> Context -> [(Symbol, Sort)] -> Expr -> Pred
+toSMT :: HasCallStack => String -> Config -> Context -> [(Symbol, Sort)] -> Expr -> Pred
 toSMT msg cfg ctx xs e =
     defuncAny cfg symenv .
         elaborate (ElabParam (solverFlags $ solver cfg) (dummyLoc msg) (elabEnv xs)) .

--- a/src/Language/Fixpoint/Solver/Common.hs
+++ b/src/Language/Fixpoint/Solver/Common.hs
@@ -15,13 +15,18 @@ import GHC.Stack (HasCallStack)
 mytracepp :: (PPrint a) => String -> a -> a
 mytracepp = notracepp
 
-askSMT :: HasCallStack => Config -> [(Symbol, Sort)] -> Expr -> SmtM Bool
-askSMT cfg xs e
---   | isContraPred e  = return False
+askSMT
+  :: HasCallStack
+  => Config
+  -> [(Symbol, Sort)] -- ^ symbols already declared in the SMT solver
+  -> [(Symbol, Sort)] -- ^ symbols to declare in the SMT solver
+  -> Expr
+  -> SmtM Bool
+askSMT cfg bsInSMT xs e
   | isTautoPred  e     = return True
   | null (kvarsExpr e) =
       do ctx <- get
-         let e' = toSMT "askSMT" cfg ctx xs e
+         let e' = toSMT "askSMT" cfg ctx (xs ++ bsInSMT) e
          checkValidWithContext xs PTrue e'
   | otherwise          = return False
 

--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -8,6 +8,7 @@
 --     2. "Reasoning about Functions", VMCAI 2018, https://ranjitjhala.github.io/static/reasoning-about-functions.pdf
 --------------------------------------------------------------------------------
 
+{-# LANGUAGE FlexibleContexts          #-}
 {-# LANGUAGE OverloadedStrings         #-}
 {-# LANGUAGE PartialTypeSignatures     #-}
 {-# LANGUAGE TupleSections             #-}
@@ -245,6 +246,7 @@ pleTrie t env = loopT env ctx0 diff0 Nothing res0 t
       , icEtaBetaFlag        = etabeta        $ ieCfg env
       , icExtensionalityFlag = extensionality $ ieCfg env
       , icLocalRewritesFlag  = localRewrites  $ ieCfg env
+      , icFreshExistentialCounter = 0
       }
 
 loopT
@@ -431,6 +433,7 @@ data ICtx    = ICtx
                                                      -- See Note [Eta expansion].
   , icExtensionalityFlag :: Bool                     -- ^ True if the extensionality flag is turned on
   , icLocalRewritesFlag  :: Bool                     -- ^ True if the local rewrites flag is turned on
+  , icFreshExistentialCounter :: Int                 -- ^ Counter to generate fresh names for existentials
   }
 
 ----------------------------------------------------------------------------------------------
@@ -482,6 +485,7 @@ updCtx InstEnv{..} ieSMT ictx delta cidMb mCTrie =
        , icANFs   = anfBinds
        , icLRWs   = mconcat $ icLRWs ictx : newLRWs
        , icBindIds = ibinds
+       , icFreshExistentialCounter = existentialCounter
        }
   where
     ibinds = insertsIBindEnv delta (icBindIds ictx)
@@ -495,7 +499,9 @@ updCtx InstEnv{..} ieSMT ictx delta cidMb mCTrie =
     es        = expr <$> bs
     eRhs      = maybe PTrue crhs subMb
 
-    binds     = [ maybeApplyKVarSolutions (x, y)
+    (binds, existentialCounter) = renameExistentialsInSortedRefts binds0 (icFreshExistentialCounter ictx)
+
+    binds0    = [ maybeApplyKVarSolutions (x, y)
                 | i <- delta
                 , let (x, y, _) = lookupBindEnv i ieBEnv
                 ]
@@ -1510,3 +1516,39 @@ checkFuel f = do
   case (M.lookup f (fcMap fc), fcMax fc) of
     (Just fk, Just n) -> pure (fk <= n)
     _                 -> pure True
+
+-- | Renames existential variables in the predicates of the given bindings to make them
+-- unique.
+--
+-- Rather than looking for all existential bindings, this function only renames
+-- the superficial existentials which can be introduced by KVar solutions.
+--
+-- These superficial existentials appear in conjunctions, disjunctions and in the
+-- body of other existentials only.
+renameExistentialsInSortedRefts
+  :: [(Symbol, SortedReft)]
+  -> Int
+  -> ([(Symbol, SortedReft)], Int)
+renameExistentialsInSortedRefts binds0 existentialCounter =
+    let
+        binds = [ (x, sr { sr_reft = mapPredReft (const p) (sr_reft sr) }) | ((x, sr), p) <- zip binds0 preds ]
+        (preds, existentialCounter') =
+          renameKVarExistentials (map (reftPred . sr_reft . snd) binds0) existentialCounter
+     in
+        (binds, existentialCounter')
+
+renameKVarExistentials :: [Expr] -> Int -> ([Expr], Int)
+renameKVarExistentials = runState . mapM go
+  where
+    go (POr es) = POr <$> mapM go es
+    go (PAnd es) = PAnd <$> mapM go es
+    go (PExist bs e0) = do
+      i1 <- get
+      let i2 = i1 + length bs
+      put i2
+      let vs = map fst bs
+          vs' = [ existSymbol v (fromIntegral i) | (v, i) <- zip vs [i1..] ]
+          bs' = zip vs' (map snd bs)
+          su = mkSubst $ zip vs (map EVar vs')
+      PExist bs' <$> go (rapierSubstExpr (S.fromList vs') su e0)
+    go e = pure e

--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -380,23 +380,25 @@ evalCandsLoop cfg ictx0 γ = go ictx0 0
         then return ictx
         else do liftSMT $ SMT.smtAssertDecl $ pAndNoDedup $ S.toList $ icAssms ictx
                 let ictx' = ictx { icAssms = mempty }
-                    (sts, candSets) = unzip $ M.toList $ icCands ictx
+                    (scopes, candSets) = unzip $ M.toList $ icCands ictx
                     cands = map S.toList candSets
-                (candss, uss) <- unzip <$>
-                  zipWithM (\s -> withExScope s . (>>= collectEqs) . mapM (evalOne γ ictx' i)) sts cands
-                let noCandidateChanged = and $ map and $ (zipWith (zipWith eqCand) candss cands)
-                    unknownEqs = M.unionWith S.difference (M.fromList (zip sts uss)) (icEquals ictx)
-                if all null unknownEqs && noCandidateChanged
-                      then return ictx
-                      else do ctx' <- gets evKCtx
-                              let eqsSMT = S.unions $ M.elems $
-                                    M.mapWithKey
-                                      (\scope -> S.map $ evalToSMT "evalCandsLoop" cfg ctx' scope)
-                                      unknownEqs
-                                  ictx'' = ictx { icEquals = M.unionWith S.union (icEquals ictx) unknownEqs
-                                                , icAssms  = S.filter (not . isTautoPred) eqsSMT
-                                                }
-                              go (ictx'' { icCands = M.fromList $ zip sts (map (S.fromList . concat) candss) }) (i + 1)
+                (candss, uss) <- unzip <$> zipWithM (evalCand ictx' i) scopes cands
+                let noCandidateChanged = all and $ zipWith (zipWith eqCand) candss cands
+                    unknownEqs = M.unionWith S.difference (M.fromList (zip scopes uss)) (icEquals ictx)
+                if all null unknownEqs && noCandidateChanged then
+                  return ictx
+                else do
+                  ctx' <- gets evKCtx
+                  let eqsSMT =
+                        S.unions $ M.elems $
+                          M.mapWithKey
+                            (\scope -> S.map $ evalToSMT "evalCandsLoop" cfg ctx' scope)
+                            unknownEqs
+                      ictx'' = ictx
+                        { icEquals = M.unionWith S.union (icEquals ictx) unknownEqs
+                        , icAssms  = S.filter (not . isTautoPred) eqsSMT
+                        }
+                  go (ictx'' { icCands = M.fromList $ zip scopes (map (S.fromList . concat) candss) }) (i + 1)
 
     testForInconsistentEnvironment :: EvalST Bool
     testForInconsistentEnvironment =
@@ -404,6 +406,9 @@ evalCandsLoop cfg ictx0 γ = go ictx0 0
 
     eqCand [e0] e1 = e0 == e1
     eqCand _ _ = False
+
+    evalCand :: ICtx -> Int -> ExScope -> [Expr] -> EvalST ([[Expr]], S.HashSet (Expr, Expr))
+    evalCand ictx i scope es = withExScope scope $ mapM (evalOne γ ictx i) es >>= collectEqs
 
     collectEqs :: [[Expr]] -> EvalST ([[Expr]], S.HashSet (Expr, Expr))
     collectEqs es = do

--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -55,7 +55,7 @@ import Language.REST.RuntimeTerm as RT
 import Language.REST.SMT (withZ3, SolverHandle)
 
 import           Control.Exception.Base (bracket)
-import           Control.Monad (filterM, foldM, forM_, when, replicateM)
+import           Control.Monad (filterM, foldM, forM_, when, replicateM, zipWithM)
 import           Control.Monad.State
 import           Control.Monad.Trans.Maybe
 import           Data.Bifunctor (second)
@@ -162,6 +162,7 @@ instEnv cfg info s cs restSolver = do
               { evEnv = SMT.ctxSymEnv ctx
               , evElabF = ef
               , evKCtx = ctx
+              , evExScope = []
               , evPendingUnfoldings = mempty
               , evNewEqualities = mempty
               , evSMTCache = mempty
@@ -247,6 +248,7 @@ pleTrie t env = loopT env ctx0 diff0 Nothing res0 t
       , icExtensionalityFlag = extensionality $ ieCfg env
       , icLocalRewritesFlag  = localRewrites  $ ieCfg env
       , icFreshExistentialCounter = 0
+      , icInitialLHSs  = mempty
       }
 
 loopT
@@ -310,12 +312,18 @@ withAssms
   -> SmtM b
 withAssms env ctx delta cidMb mCTrie act = do
   sctx <- get
-  let ictx' = updCtx env sctx ctx delta cidMb mCTrie
+  let (ictx', bs) = updCtx env sctx ctx delta cidMb mCTrie
   let assms = icAssms ictx'
 
   SMT.smtBracket "PLE.withAssms" $ do
-    forM_ assms SMT.smtAssertDecl
+    -- See Note [Existential quantification when unfolding]
+    SMT.smtDecls $ elabBindings (ieEvEnv env) bs
+    forM_ (S.toList assms) SMT.smtAssertDecl
     act $ ictx' { icAssms = mempty }
+
+  where
+    elabBindings eenv bs =
+      elaborate (ElabParam (evElabF eenv) "withAssms: PExist Args" (evEnv eenv)) bs
 
 -- | @ple1@ performs the PLE at a single "node" in the Trie
 --
@@ -328,17 +336,23 @@ ple1 ie@InstEnv{..} ictx i res = do
   (ictx', env) <- liftIO $ runStateT (evalCandsLoop ieCfg ictx ieKnowl) (ieEvEnv { evKCtx = ctx })
   put $ evKCtx env
   let pendings = collectPendingUnfoldings env (icSubcId ictx)
-      newEqs = pendings ++ S.toList (S.difference (icEquals ictx') (icEquals ictx))
+      newEqs =
+        reconstructExistentials
+          (M.intersectionWith S.union (icInitialLHSs ictx) $         -- add original predicates
+           M.map (S.map equalitiesPred) $                            -- construct equalities
+           M.unionWith S.union pendings $                            -- pending unfoldings if any
+           M.unionWith S.difference (icEquals ictx') (icEquals ictx) -- new equalities only
+          )
   return (ictx', ie { ieEvEnv = env }, updCtxRes res i newEqs)
   where
     -- Pending unfoldings (i.e. with undecided guards) are collected only
     -- when we reach a leaf in the Trie, and only if the user asked for them.
     collectPendingUnfoldings env (Just _) | pleUndecGuards ieCfg =
-      M.toList (evPendingUnfoldings env)
-    collectPendingUnfoldings _ _ = []
+      M.map (S.fromList . M.toList) (evPendingUnfoldings env)
+    collectPendingUnfoldings _ _ = mempty
 
-evalToSMT :: String -> Config -> SMT.Context -> (Expr, Expr) -> Pred
-evalToSMT msg cfg ctx (e1,e2) = toSMT ("evalToSMT:" ++ msg) cfg ctx [] (EEq e1 e2)
+evalToSMT :: String -> Config -> SMT.Context -> [(Symbol, Sort)] -> (Expr, Expr) -> Pred
+evalToSMT msg cfg ctx bs (e1,e2) = toSMT ("evalToSMT:" ++ msg) cfg ctx bs (EEq e1 e2)
 
 -- | Generate equalities for all function invocations in the candidates
 -- in @ctx@ for which definitions are known. The function definitions are in
@@ -359,33 +373,53 @@ evalCandsLoop :: Config -> ICtx -> Knowledge -> EvalST ICtx
 evalCandsLoop cfg ictx0 γ = go ictx0 0
   where
     go :: ICtx -> Int -> EvalST ICtx
-    go ictx _ | S.null (icCands ictx) = return ictx
+    go ictx _ | all null (icCands ictx) = return ictx
     go ictx i = do
       inconsistentEnv <- testForInconsistentEnvironment
       if inconsistentEnv
         then return ictx
-        else do liftSMT $ SMT.smtAssertDecl (pAndNoDedup (S.toList $ icAssms ictx))
+        else do liftSMT $ SMT.smtAssertDecl $ pAndNoDedup $ S.toList $ icAssms ictx
                 let ictx' = ictx { icAssms = mempty }
-                    cands = S.toList $ icCands ictx
-                candss <- mapM (evalOne γ ictx' i) cands
-                us <- gets evNewEqualities
-                modify $ \st -> st { evNewEqualities = mempty }
-                let noCandidateChanged = and (zipWith eqCand candss cands)
-                    unknownEqs = us `S.difference` icEquals ictx
-                if S.null unknownEqs && noCandidateChanged
+                    (sts, candSets) = unzip $ M.toList $ icCands ictx
+                    cands = map S.toList candSets
+                (candss, uss) <- unzip <$>
+                  zipWithM (\s -> withExScope s . (>>= collectEqs) . mapM (evalOne γ ictx' i)) sts cands
+                let noCandidateChanged = and $ map and $ (zipWith (zipWith eqCand) candss cands)
+                    unknownEqs = M.unionWith S.difference (M.fromList (zip sts uss)) (icEquals ictx)
+                if all null unknownEqs && noCandidateChanged
                       then return ictx
                       else do ctx' <- gets evKCtx
-                              let eqsSMT = evalToSMT "evalCandsLoop" cfg ctx' `S.map` unknownEqs
-                                  ictx'' = ictx { icEquals = icEquals ictx <> unknownEqs
-                                                 , icAssms  = S.filter (not . isTautoPred) eqsSMT }
-                              go (ictx'' { icCands = S.fromList (concat candss) }) (i + 1)
+                              let eqsSMT = S.unions $ M.elems $
+                                    M.mapWithKey
+                                      (\scope -> S.map $ evalToSMT "evalCandsLoop" cfg ctx' scope)
+                                      unknownEqs
+                                  ictx'' = ictx { icEquals = M.unionWith S.union (icEquals ictx) unknownEqs
+                                                , icAssms  = S.filter (not . isTautoPred) eqsSMT
+                                                }
+                              go (ictx'' { icCands = M.fromList $ zip sts (map (S.fromList . concat) candss) }) (i + 1)
 
     testForInconsistentEnvironment :: EvalST Bool
     testForInconsistentEnvironment =
-      liftSMT $ knPreds γ (knLams γ) PFalse
+      knPredsEvalST γ PFalse
 
     eqCand [e0] e1 = e0 == e1
     eqCand _ _ = False
+
+    collectEqs :: [[Expr]] -> EvalST ([[Expr]], S.HashSet (Expr, Expr))
+    collectEqs es = do
+      env <- get
+      let newEqs = evNewEqualities env
+      modify $ \st -> st { evNewEqualities = mempty }
+      return (es, newEqs)
+
+    withExScope :: ExScope -> EvalST a -> EvalST a
+    withExScope s m = do
+      env <- get
+      put $ env { evExScope = s }
+      r <- m
+      modify $ \st -> st { evExScope = evExScope env }
+      return r
+
 
 ----------------------------------------------------------------------------------------------
 -- | Step 3: @resSInfo@ uses incremental PLE result @InstRes@ to produce the strengthened SInfo
@@ -420,8 +454,8 @@ data InstEnv a = InstEnv
 
 data ICtx    = ICtx
   { icAssms              :: S.HashSet Pred           -- ^ Equalities converted to SMT format
-  , icCands              :: S.HashSet Expr           -- ^ "Candidates" for unfolding
-  , icEquals             :: EvEqualities             -- ^ Accumulated equalities
+  , icCands              :: M.HashMap ExScope (S.HashSet Expr)  -- ^ "Candidates" for unfolding
+  , icEquals             :: M.HashMap ExScope EvEqualities      -- ^ Accumulated equalities
   , icSimpl              :: !ConstMap                -- ^ Map of expressions to constants
   , icSubcId             :: Maybe SubcId             -- ^ Current subconstraint ID
   , icANFs               :: [[(Symbol, SortedReft)]] -- Hopefully contain only ANF things
@@ -434,6 +468,8 @@ data ICtx    = ICtx
   , icExtensionalityFlag :: Bool                     -- ^ True if the extensionality flag is turned on
   , icLocalRewritesFlag  :: Bool                     -- ^ True if the local rewrites flag is turned on
   , icFreshExistentialCounter :: Int                 -- ^ Counter to generate fresh names for existentials
+  , icInitialLHSs :: M.HashMap ExScope (S.HashSet Expr)
+                                                     -- ^ LHS candidates before any unfoldings
   }
 
 ----------------------------------------------------------------------------------------------
@@ -452,11 +488,13 @@ type CTrie   = T.Trie   SubcId
 type CBranch = T.Branch SubcId
 type Diff    = [BindId]    -- ^ in "reverse" order
 
-equalitiesPred :: [(Expr, Expr)] -> [Expr]
-equalitiesPred eqs = [ EEq e1 e2 | (e1, e2) <- eqs, e1 /= e2 ]
+equalitiesPred :: (Expr, Expr) -> Expr
+equalitiesPred (e1, e2)
+  | e1 /= e2 = EEq e1 e2
+  | otherwise = PTrue
 
-updCtxRes :: InstRes -> Maybe BindId -> [(Expr, Expr)] -> InstRes
-updCtxRes res iMb = updRes res iMb . pAndNoDedup . equalitiesPred
+updCtxRes :: InstRes -> Maybe BindId -> [Expr] -> InstRes
+updCtxRes res iMb = updRes res iMb . pAndNoDedup
 
 
 updRes :: InstRes -> Maybe BindId -> Expr -> InstRes
@@ -466,6 +504,9 @@ updRes res  Nothing _ = res
 ----------------------------------------------------------------------------------------------
 -- | @updCtx env ctx delta cidMb@ adds the assumptions and candidates from @delta@ and @cidMb@
 --   to the context.
+--
+-- Yields the new context and a list of existential binders found in @delta@.
+-- See Note [Existential quantification when unfolding].
 ----------------------------------------------------------------------------------------------
 
 updCtx
@@ -476,24 +517,27 @@ updCtx
   -> Diff
   -> Maybe SubcId
   -> Maybe CTrie
-  -> ICtx
+  -> (ICtx, [(Symbol, Sort)])
 updCtx InstEnv{..} ieSMT ictx delta cidMb mCTrie =
-  ictx { icAssms  = S.fromList (filter (not . isTautoPred) ctxEqs)
-       , icCands  = S.fromList deANFedCands <> icCands ictx
-       , icSimpl  = icSimpl ictx <> econsts
-       , icSubcId = cidMb
-       , icANFs   = anfBinds
-       , icLRWs   = mconcat $ icLRWs ictx : newLRWs
-       , icBindIds = ibinds
-       , icFreshExistentialCounter = existentialCounter
-       }
+    ( ictx { icAssms  = S.fromList ctxEqs
+           , icCands  = M.unionWith S.union candsPerExScope (icCands ictx)
+           , icSimpl  = icSimpl ictx <> econsts
+           , icSubcId = cidMb
+           , icANFs   = anfBinds
+           , icLRWs   = mconcat $ icLRWs ictx : newLRWs
+           , icBindIds = ibinds
+           , icFreshExistentialCounter = existentialCounter
+           , icInitialLHSs = M.unionWith S.union candsPerExScopeNoRHS (icInitialLHSs ictx)
+           }
+    , S.toList $ S.fromList $ concat $ M.keys candsPerExScope
+    )
   where
     ibinds = insertsIBindEnv delta (icBindIds ictx)
     cands     = rhs:es
     anfBinds  = bs : icANFs ictx
     econsts   = M.fromList $ findConstants ieKnowl es
-    ctxEqs    = toSMT "updCtx" ieCfg ieSMT [] <$> L.nub
-                  [ c | xr <- bs, c <- conjuncts (expr xr), null (Vis.kvarsExpr c) ]
+    ctxEqs    = fmap (toSMT "updCtx" ieCfg ieSMT []) $ L.nub $ filter (null . Vis.kvarsExpr)
+                  [ c | xr <- bs, c <- conjuncts (expr xr), not (isTautoPred c) ]
     bs        = second unApplySortedReft <$> binds
     rhs       = unApply eRhs
     es        = expr <$> bs
@@ -508,7 +552,10 @@ updCtx InstEnv{..} ieSMT ictx delta cidMb mCTrie =
     subMb     = getCstr ieCstrs <$> cidMb
     newLRWs   = Mb.mapMaybe (`lookupLocalRewrites` ieLRWs) delta
 
-    deANFedCands =
+    candsPerExScopeNoRHS = M.fromListWith S.union $ ([], S.empty) : drop 1 deANFedCands
+    candsPerExScope = M.unionWith S.union candsPerExScopeNoRHS $ M.fromListWith S.union (take 1 deANFedCands)
+
+    deANFedCands = map (second S.singleton . prenexExistentials) $
       -- We only call 'deANF' if necessary.
       if not (null (getAutoRws ieKnowl cidMb))
          || icExtensionalityFlag ictx
@@ -559,8 +606,14 @@ data EvalEnv = EvalEnv
   { evEnv      :: !SymEnv
   , evElabF    :: ElabFlags
   , evKCtx     :: SMT.Context
-    -- | Equalities where we couldn't evaluate the guards
-  , evPendingUnfoldings :: M.HashMap Expr Expr
+    -- | The current scope of existential variables.
+    -- See Note [Existential quantification when unfolding].
+  , evExScope  :: ExScope
+    -- | Equalities where we couldn't evaluate the guards, in a map which
+    -- uses their existential scope as key.
+    --
+    -- See Note [Existential quantification when unfolding].
+  , evPendingUnfoldings :: M.HashMap ExScope (M.HashMap Expr Expr)
   , evNewEqualities :: EvEqualities -- ^ Equalities discovered during a traversal of
                                     -- an expression
   , evSMTCache :: M.HashMap Expr Bool -- ^ Whether an expression is valid or its negation
@@ -837,13 +890,14 @@ evalRESTWithCache cacheRef _ ctx acc rp
 evalRESTWithCache cacheRef γ ctx acc rp =
   do
     mexploredTerms <- gets explored
+    ebs <- gets evExScope
     case mexploredTerms of
       Nothing -> return acc
       Just exploredTerms -> do
-        se <- liftIO (shouldExploreTerm exploredTerms exprs)
+        se <- liftIO (shouldExploreTerm ebs exploredTerms exprs)
         if se then do
-          possibleRWs <- liftSMT getRWs
-          rws <- notVisitedFirst exploredTerms <$> filterM (liftIO . allowed) possibleRWs
+          possibleRWs <- liftSMT (getRWs ebs)
+          rws <- notVisitedFirst exploredTerms <$> filterM (liftIO . allowed ebs) possibleRWs
           oldEqualities <- gets evNewEqualities
           modify $ \st -> st { evNewEqualities = mempty }
 
@@ -885,17 +939,17 @@ evalRESTWithCache cacheRef γ ctx acc rp =
         else
           return acc
   where
-    shouldExploreTerm exploredTerms e | Vis.isConc e =
-      case rwTerminationOpts rwArgs of
+    shouldExploreTerm ebs exploredTerms e | Vis.isConc e =
+      case rwTerminationOpts (rwArgs ebs) of
         RWTerminationCheckDisabled ->
           return $ not $ ExploredTerms.visited (Rewrite.convert e) exploredTerms
         RWTerminationCheckEnabled  ->
           ExploredTerms.shouldExplore (Rewrite.convert e) (c rp) exploredTerms
-    shouldExploreTerm _ _ = return False
+    shouldExploreTerm _ _ _ = return False
 
-    allowed (_, rwE, _) | rwE `elem` pathExprs = return False
-    allowed (_, _, c)   = termCheck c
-    termCheck c = Rewrite.passesTerminationCheck (oc rp) rwArgs c
+    allowed _ebs (_, rwE, _) | rwE `elem` pathExprs = return False
+    allowed ebs (_, _, c)   = termCheck ebs c
+    termCheck ebs c = Rewrite.passesTerminationCheck (oc rp) (rwArgs ebs) c
 
     notVisitedFirst exploredTerms rws =
       let
@@ -921,20 +975,20 @@ evalRESTWithCache cacheRef γ ctx acc rp =
     exprs           = last pathExprs
     autorws         = getAutoRws γ (icSubcId ctx)
 
-    rwArgs = RWArgs (isValid cacheRef γ) $ knRWTerminationOpts γ
+    rwArgs ebs = RWArgs (isValid cacheRef ebs γ) $ knRWTerminationOpts γ
 
-    getRWs =
+    getRWs ebs =
       do
         -- Optimization: If we got here via rewriting, then the current constraints
         -- are satisfiable; otherwise double-check that rewriting is still allowed
         ok <-
           if isRW $ last (path rp)
             then return True
-            else liftIO $ termCheck (c rp)
+            else liftIO $ termCheck ebs (c rp)
         if ok
           then
             do
-              let getRW e ar = Rewrite.getRewrite (oc rp) rwArgs (c rp) e ar
+              let getRW e ar = Rewrite.getRewrite (oc rp) (rwArgs ebs) (c rp) e ar
               let getRWs' s  = Mb.catMaybes <$> mapM (runMaybeT . getRW s) autorws
               concat <$> mapM getRWs' (subExprs exprs)
           else return []
@@ -1014,14 +1068,15 @@ evalApp γ ctx e0 es et
            -- If evalIte does any modifications, though, we do unfold in order
            -- to allow analysis of the resulting expression
            modify $ \st -> st
-             { evPendingUnfoldings = M.insert (eApps e0 es) e3' (evPendingUnfoldings st)
+             { evPendingUnfoldings =
+                 M.insertWith M.union (evExScope st) (M.singleton (eApps e0 es) e3') (evPendingUnfoldings st)
              }
            return Nothing
          else do
            useFuel f
            modify $ \st -> st
              { evNewEqualities = S.insert (eApps e0 es, e3') (evNewEqualities st)
-             , evPendingUnfoldings = M.delete (eApps e0 es) (evPendingUnfoldings st)
+             , evPendingUnfoldings = M.adjust (M.delete (eApps e0 es)) (evExScope st) (evPendingUnfoldings st)
              }
            return (Just $ eApps e2' es2)
        else return Nothing
@@ -1221,14 +1276,14 @@ isValidCached γ e = do
   case M.lookup e (evSMTCache env) of
     Nothing -> do
       let isFreeInE (s, _) = not (S.member s (exprSymbolsSet e))
-      b <- liftSMT $ knPreds γ (knLams γ) e
+      b <- knPredsEvalST γ e
       if b
         then do
           when (all isFreeInE (knLams γ)) $
             put (env { evSMTCache = M.insert e True (evSMTCache env) })
           return (Just True)
         else do
-          b2 <- liftSMT $ knPreds γ (knLams γ) (PNot e)
+          b2 <- knPredsEvalST γ (PNot e)
           if b2
             then do
               when (all isFreeInE (knLams γ)) $
@@ -1250,7 +1305,10 @@ data Knowledge = KN
     -- user data declaration.
     knSims              :: Map Symbol [(Rewrite, IsUserDataSMeasure)]
   , knAms               :: Map Symbol Equation -- ^ All function definitions
-  , knPreds             :: [(Symbol, Sort)] -> Expr -> SmtM Bool
+    -- | @knPreds γ bsInSMT xs e@ checks whether @e@ is valid under the
+    -- assumptions that all variables in @bsInSMT@ are in the SMT solver,
+    -- and that all variables in @xs@ need tp be declared in the SMT solver.
+  , knPreds             :: [(Symbol, Sort)] -> [(Symbol, Sort)] -> Expr -> SmtM Bool
   , knLams              :: ![(Symbol, Sort)]
   , knSummary           :: ![(Symbol, Int)]     -- ^ summary of functions to be evaluates (knSims and knAsms) with their arity
   , knDCs               :: !(S.HashSet Symbol)  -- ^ data constructors drawn from Rewrite
@@ -1266,12 +1324,17 @@ data Knowledge = KN
 data IsUserDataSMeasure = NoUserDataSMeasure | UserDataSMeasure
   deriving (Eq, Show)
 
-isValid :: IORef (M.HashMap Expr Bool) -> Knowledge -> Expr -> SmtM Bool
-isValid cacheRef γ e = do
+knPredsEvalST :: Knowledge -> Expr -> EvalST Bool
+knPredsEvalST γ e = do
+    env <- get
+    liftSMT $ knPreds γ (evExScope env) (knLams γ) e
+
+isValid :: IORef (M.HashMap Expr Bool) -> [(Symbol, Sort)] -> Knowledge -> Expr -> SmtM Bool
+isValid cacheRef bs γ e = do
     smtCache <- liftIO $ readIORef cacheRef
     case M.lookup e smtCache of
       Nothing -> do
-        b <- knPreds γ (knLams γ) e
+        b <- knPreds γ bs (knLams γ) e
         when b $
           liftIO $ writeIORef cacheRef (M.insert e True smtCache)
         return b
@@ -1504,7 +1567,8 @@ makeFreshEtaNames n = replicateM n makeFreshName
 elaborateExpr :: String -> Expr -> EvalST Expr
 elaborateExpr msg e = do
   let elabSpan = atLoc dummySpan msg
-  symEnv' <- gets evEnv
+  env <- get
+  let symEnv' = insertsSymEnv (evEnv env) (evExScope env)
   ef <- gets evElabF
   pure $ unApply $ elaborate (ElabParam ef elabSpan symEnv') e
 
@@ -1517,8 +1581,58 @@ checkFuel f = do
     (Just fk, Just n) -> pure (fk <= n)
     _                 -> pure True
 
--- | Renames existential variables in the predicates of the given bindings to make them
--- unique.
+
+-- Note [Existential quantification when unfolding]
+-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+--
+-- After FUSION is performed, some predicates, which previously used kvars, may
+-- contain existential quantifications.
+--
+-- When the unfoldings are searched by PLE in expressions with existentials,
+-- we make sure that the produced unfoldings still have the existential
+-- bindings in scope.
+--
+-- The procedure is as follows:
+-- 1. First, we rename the existential variables in the predicates of the bindings
+--    to make them unique ('renameExistentialsInSortedRefts').
+--
+--    @exists x y. f x y || (exists x. g x y)@
+--
+--    becomes
+--
+--    @exists v0 v1. f v0 v1 || (exists v2. g v2 v1)@
+--
+-- 2. We extract the nested existentials to prenex form, and we store the bodies of
+--    the existentials in a map with the existential binders as keys
+--    ('prenexExistentials' and 'updCtx').
+--
+--    @exists v0 v1. f v0 v1 || (exists v2. g v2 v1)@
+--
+--    produces the map
+--
+--    @[v0, v1, v2] -> f v0 v1 || g v2 v1@
+--
+-- 3. We declare to the SMT solver the existential variables in every scope
+--    (in 'withAssms').
+--
+-- 4. We then look for unfoldings in each of the subexpressions. Whenever
+--    we find an unfolding, we record the scope in which it was found.
+--
+--    @[v0, v1, v3] -> (f v0 v1 = v0 < v1) && (g v2 = v2 > v1)@
+--
+-- 5. When PLE is finished, we create for every scope an existential
+--    quantification whose body contains all the corresponding unfoldings
+--    and the original subexpressions in the scope ('reconstructExistentials').
+--
+--    @exists v0 v1 v0.
+--       (f v0 v 1 = v0 < v1) && (g v2 = v2 > v1) &&
+--       (f v0 v1 || g v2 v1)@
+--
+--    This is the expression that PLE returns.
+
+
+-- | Renames existential variables in the predicates of the given bindings to
+-- make them unique.
 --
 -- Rather than looking for all existential bindings, this function only renames
 -- the superficial existentials which can be introduced by KVar solutions.
@@ -1552,3 +1666,42 @@ renameKVarExistentials = runState . mapM go
           su = mkSubst $ zip vs (map EVar vs')
       PExist bs' <$> go (rapierSubstExpr (S.fromList vs') su e0)
     go e = pure e
+
+-- ^ Scopes of existential binders identifying the location of sub-expressions
+type ExScope = [(Symbol, Sort)]
+
+
+-- | Extracts nested existentials from an expression.
+--
+-- For example, the expression
+--
+-- > exists [x1 : t1]. e1 == e2 &&
+-- > exists [x2 : t2]. e3 == 2 &&
+-- > exists [x3 : t3]. e3 < e4
+--
+-- would be flattened into
+--
+-- > (e1 == e2 && e3 == 2 && e3 < e4, [x1 : t1, x2 : t2, x3 : t3])
+--
+-- Precondition: the existential binding names are unique.
+--
+prenexExistentials :: Expr -> (ExScope, Expr)
+prenexExistentials = go
+  where
+    go :: Expr -> (ExScope, Expr)
+    go (PExist bs e) =
+      let (bs', e') = go e
+      in (bs ++ bs', e')
+    go (PAnd es) =
+      let (bss, es') = unzip (map go es)
+      in (concat bss, PAnd es')
+    go (POr es) =
+      let (bss, es') = unzip (map go es)
+      in (concat bss, POr es')
+    go e = ([], e)
+
+
+-- | Reconstructs expressions with existentials from a map
+-- of existential scopes to their bodies.
+reconstructExistentials :: M.HashMap ExScope (S.HashSet Expr) -> [Expr]
+reconstructExistentials m = [ pExist s (pAndNoDedup $ S.toList es) | (s, es) <- M.toList m, not (null es) ]

--- a/src/Language/Fixpoint/SortCheck.hs
+++ b/src/Language/Fixpoint/SortCheck.hs
@@ -137,7 +137,7 @@ data ElabParam = ElabParam
   }
 
 class Elaborate a where
-  elaborate :: ElabParam -> a -> a
+  elaborate :: HasCallStack => ElabParam -> a -> a
 
 
 instance (Loc a) => Elaborate (SInfo a) where
@@ -203,7 +203,7 @@ elabDefinedEqn ep eq = eq { eqBody = elaborateExpr ep' (eqBody eq) (Just t')
 instance Elaborate Expr where
   elaborate p e = elaborateExpr p e Nothing
 
-elaborateExpr :: ElabParam -> Expr -> Maybe Sort -> Expr
+elaborateExpr :: HasCallStack => ElabParam -> Expr -> Maybe Sort -> Expr
 elaborateExpr (ElabParam ef msg env) e t =
   elabNumeric . elabApply env' . elabExpr (ElabParam ef msg env') t .  elabSorts ef . elabFMap . (if Cfg.elabSetBag ef then elabFSetBagZ3 else id) $ e
     where
@@ -436,7 +436,7 @@ elabSorts _ e                 = e
 --------------------------------------------------------------------------------
 -- | 'elabExpr' adds "casts" to decorate polymorphic instantiation sites.
 --------------------------------------------------------------------------------
-elabExpr :: ElabParam -> Maybe Sort -> Expr -> Expr
+elabExpr :: HasCallStack => ElabParam -> Maybe Sort -> Expr -> Expr
 elabExpr ep t e = case elabExprE ep t e of
   Left ex  -> die ex
   Right e' -> F.notracepp ("elabExp " ++ showpp e) e'

--- a/src/Language/Fixpoint/Types/Errors.hs
+++ b/src/Language/Fixpoint/Types/Errors.hs
@@ -138,7 +138,7 @@ err :: SrcSpan -> Doc -> Error
 err sp d = Error [Error1 sp d]
 
 ---------------------------------------------------------------------
-panic :: String -> a
+panic :: HasCallStack => String -> a
 ---------------------------------------------------------------------
 panic = die . err dummySpan . text . (panicMsg ++)
 

--- a/src/Language/Fixpoint/Types/Names.hs
+++ b/src/Language/Fixpoint/Types/Names.hs
@@ -521,26 +521,42 @@ hvarPrefix  = "nnf_arg$"
 unKArgSymbol :: Symbol -> Symbol
 unKArgSymbol = unSuffixSymbol . unSuffixSymbol . unPrefixSymbol kArgPrefix
 
--- | 'tidySymbol' is used to prettify the names of parameters of kvars appearing in solutions.(*)
---   For example, if you have a kvar $k0 with two parameters, you may have a solution that looks like
---       0 <  lq_karg$nnf_arg$##k0##0##k0
+-- | @tidySymbol@ is used to prettify the names of parameters of kvars appearing
+-- in solutions.  For example, if you have a kvar $k0 with two parameters, you
+-- may have a solution that looks like
+--
+-- > 0 <  lq_karg$nnf_arg$##k0##0##k0
+--
 --   where we know it is a kvar-arg because of the
---      - `kArgPrefix` (`lq_arg`)
---      - `hvarArgPrefix` (`nnf_arg`)
---      - `k0` the name of the kvar
---      - `0`  the parameter index
---      - `k0` again (IDK why?!)
---    all of which are separated by `##`
---   So `tidySymbol` tests if indeed it is a `kArgPrefix`-ed symbol and if so converts
---      `lq_karg$nnf_arg$##k0##0##k0` ----> `$k0##0`
+--      - @kArgPrefix@ (@lq_arg@)
+--      - @hvarArgPrefix@ (@nnf_arg@)
+--      - @k0@ the name of the kvar
+--      - @0@  the parameter index
+--      - @k0@ again (IDK why?!)
+--    all of which are separated by @##@
+--
+--   So @tidySymbol@ tests if indeed it is a @kArgPrefix@-ed symbol and if so
+--   converts
+--
+-- > lq_karg$nnf_arg$##k0##0##k0 ----> $k0##0
+--
+--  KArgs from Liquid Haskell come in the form @k_##0@ instead, and parameters
+--  are like @lq_karg$param_name##0##k_##0@. In this case, tidySymbol will
+--  convert
+--
+--  > lq_karg$param_name##0##k_##0  ----> $param_name##0##k_
 
 tidySymbol :: Symbol -> Symbol
 tidySymbol s
   | s == s'   = s
   | otherwise = s''
   where
-    s'        = unPrefixSymbol kArgPrefix s
-    s''       = consSym '$' . unPrefixSymbol symSepName . unSuffixSymbol . unPrefixSymbol hvarPrefix $ s'
+    s' = unPrefixSymbol kArgPrefix s
+    s'' =
+      consSym '$' $
+      unPrefixSymbol symSepName $
+      unSuffixSymbol $
+      unPrefixSymbol hvarPrefix s'
 
 unPrefixSymbol :: Symbol -> Symbol -> Symbol
 unPrefixSymbol p s = fromMaybe s (stripPrefix p s)

--- a/tests/elim/test780.fq
+++ b/tests/elim/test780.fq
@@ -1,0 +1,32 @@
+// test that PLE can unfold in the solution of kvars (#780)
+fixpoint "--rewrite"
+
+constant adder: (func(0, [int; int; int]))
+
+define adder(x : int, y : int) : int = { x + y }
+
+expand [1 : True; 2 : True; 3 : True]
+
+bind 0 a : {v: int | $k0}
+
+constraint:
+  env [ ]
+  lhs {v : int | adder v 1 = 11}
+  rhs {v : int | $k0}
+  id 1 tag []
+
+constraint:
+  env [ ]
+  lhs {v : int | adder v 1 = 21}
+  rhs {v : int | $k0}
+  id 2 tag []
+
+constraint:
+  env [ 0 ]
+  lhs {v : int | v = a}
+  rhs {v : int | 10 <= v}
+  id 3 tag []
+
+wf:
+  env [ ]
+  reft {v: int | $k0}

--- a/tests/tasty/SimplifyPLE.hs
+++ b/tests/tasty/SimplifyPLE.hs
@@ -32,8 +32,8 @@ simplify' = PLE.simplify emptyKnowledge emptyICtx
     emptyICtx =
       ICtx
         { icAssms = S.empty,      -- S.HashSet Pred
-          icCands = S.empty,      -- :: S.HashSet Expr
-          icEquals = S.empty,     -- :: EvAccum
+          icCands = mempty,      -- :: S.HashSet Expr
+          icEquals = mempty,     -- :: EvAccum
           icSimpl = SM.empty,     -- :: !ConstMap
           icSubcId = Nothing,     -- :: Maybe SubcId
           icANFs = [],            -- :: [[(Symbol, SortedReft)]]
@@ -41,5 +41,7 @@ simplify' = PLE.simplify emptyKnowledge emptyICtx
           icEtaBetaFlag        = False,
           icExtensionalityFlag = False,
           icLocalRewritesFlag  = False,
-          icBindIds = mempty
+          icBindIds = mempty,
+          icFreshExistentialCounter = 0,
+          icInitialLHSs = mempty
         }

